packages/shared/src/components/buttons/ButtonV2.tsx

```diff
@@ -0,0 +1,193 @@
+  [ButtonSize.Small]: IconSize.Small,
+  [ButtonSize.XSmall]: IconSize.XSmall,
+};
+
+export type IconType = React.ReactElement<IconProps>;
+
+export { ButtonColor };
+
+export enum ButtonVariant {
+  Primary = 'primary',
+  Secondary = 'secondary',
+  Tertiary = 'tertiary',
+  Float = 'tertiaryFloat',
+}
+
+export enum ButtonIconPosition {
+  Left = 'left',
+  Right = 'right',
+}
+interface CommonButtonProps {
+  size?: ButtonSize;
+  loading?: boolean;
+  pressed?: boolean;
+  disabled?: boolean;
+  children?: ReactNode;
+  tag?: React.ElementType & AllowedTags;
+}
+
+// when color is present, variant is required
+type ColorButtonProps =
+  | { color: ButtonColor; variant: ButtonVariant }
+  | { color?: never; variant?: ButtonVariant };
+
+// when iconPosition is present, icon is required
+type IconButtonProps =
+  | { iconPosition: ButtonIconPosition; icon: IconType }
+  | { iconPosition?: never; icon?: IconType };
+
+type BaseButtonProps = CommonButtonProps & ColorButtonProps & IconButtonProps;
+
+const useGetIconWithSize = (
+  size: ButtonSize,
+  iconOnly: boolean,
+  iconPosition: ButtonIconPosition,
+) => {
+  return (icon: React.ReactElement<IconProps>) =>
+    React.cloneElement(icon, {
+      size: icon.props?.size ?? buttonSizeToIconSize[size],
+      className: classNames(
+        icon.props.className,
+        !iconOnly && 'text-base !w-6 !h-6',
+        !iconOnly && iconPosition === ButtonIconPosition.Left && '-ml-2 mr-1',
+        !iconOnly && iconPosition === ButtonIconPosition.Right && 'ml-1 -mr-2',
+        // add margin to social icons
+        icon.props.className?.split(' ').includes('socialIcon') && 'mr-3',
+      ),
+    });
+};
+
+export type AllowedTags = keyof Pick<JSX.IntrinsicElements, 'a' | 'button'>;
+export type AllowedElements = HTMLButtonElement | HTMLAnchorElement;
+export type ButtonElementType<Tag extends AllowedTags> = Tag extends 'a'
+  ? HTMLAnchorElement
+  : HTMLButtonElement;
+
+export type ButtonProps<T extends AllowedTags> = BaseButtonProps &
+  HTMLAttributes<T> &
+  JSX.IntrinsicElements[T] & {
+    ref?: Ref<ButtonElementType<T>>;
+  };
+
+const sizeToClassName: Record<ButtonSize, string> = {
+  [ButtonSize.XLarge]: 'border-none',
+  [ButtonSize.Large]: 'h-12 px-6 rounded-[0.88rem]',
+  [ButtonSize.Medium]: 'h-10 px-5 rounded-xl',
+  [ButtonSize.Small]: 'h-8 px-3 rounded-[0.63rem]',
+  [ButtonSize.XSmall]: 'h-6 px-2 rounded-lg',
+};
+
+const iconOnlySizeToClassName: Record<ButtonSize, string> = {
+  [ButtonSize.XLarge]: 'h-16 w-16 p-0 rounded-[1.375rem]',
+  [ButtonSize.Large]: 'h-12 w-12 p-0 rounded-[0.88rem]',
+  [ButtonSize.Medium]: 'h-10 w-10 p-0 rounded-xl',
+  [ButtonSize.Small]: 'h-8 w-8 p-0 rounded-[0.63rem]',
+  [ButtonSize.XSmall]: 'h-6 w-6 p-0 rounded-lg',
+};
+
+const variantToClassName: Record<ButtonVariant, string> = Object.values(
+  ButtonVariant,
+).reduce((acc, variant) => {
+  acc[variant] = `btn-${variant}`;
+  return acc;
+}, {} as Record<ButtonVariant, string>);
```

Comment:
I'm still not a big fan of this, it goes against tailwinds own guidelines and we generally always avoid it.
It's maybe better to consider a mapping array? (I think that's the proposal here, but it's still concatenated)

Although mostly it should probably work, tailwind might also change the JIT compiler, then this will break as it can't pre-render the strings for instance.

Reply:
I can give it another go and avoid the string concatenation 👍